crates/core/src/metadata_store/providers/etcd.rs

```diff
@@ -0,0 +1,247 @@
+
+        Ok(Self { client })
+    }
+
+    /// deletes a key only if and only if the current value in store
+    /// has the exact given version
+    async fn delete_if_has_version(
+        &self,
+        client: &mut KvClient,
+        key: Bytes,
+        version: i64,
+    ) -> Result<bool, WriteError> {
+        let txn = Txn::new()
+            .when(vec![Compare::version(
+                key.clone(),
+                CompareOp::Equal,
+                version,
+            )])
+            .and_then(vec![TxnOp::delete(key, None)]);
+
+        let response = client.txn(txn).await?;
+
+        Ok(response.succeeded())
+    }
+
+    /// puts a key/value if and only if the current value in store
+    /// has the exact given version
+    async fn put_if_has_version(
+        &self,
+        client: &mut KvClient,
+        key: Bytes,
+        value: Bytes,
+        version: i64,
+    ) -> Result<bool, WriteError> {
+        let txn = Txn::new()
+            .when(vec![Compare::version(
+                key.clone(),
+                CompareOp::Equal,
+                version,
+            )])
+            .and_then(vec![TxnOp::put(key, value, None)]);
+
+        let response = client.txn(txn).await?;
+
+        Ok(response.succeeded())
+    }
+}
+
+#[async_trait::async_trait]
+impl MetadataStore for EtcdMetadataStore {
+    async fn get(&self, key: ByteString) -> Result<Option<VersionedValue>, ReadError> {
+        let mut client = self.client.kv_client();
+        let mut response = client.get(key.into_bytes(), None).await?;
+
+        // return first value because this is supposed to be an exact match
+        // not a scan
+        let kv = match response.take_kvs().into_iter().next() {
+            None => return Ok(None),
+            Some(kv) => kv,
+        };
+
+        // please read todo! on implementation of .to_version()
+        let version = (&kv).to_version()?;
+        let (_, value) = kv.into_key_value();
+
+        if value.is_empty() {
+            // we keep an empty value in place of
+            // deleted keys to work around the version limitation
+            return Ok(None);
+        }
+
+        Ok(Some(VersionedValue::new(version, value.into())))
+    }
+
+    async fn get_version(&self, key: ByteString) -> Result<Option<Version>, ReadError> {
+        let mut client = self.client.kv_client();
+        let mut response = client.get(key.into_bytes(), None).await?;
+
+        // return first value because this suppose to be an exact match
+        // not a scan
+        let kv = match response.take_kvs().into_iter().next() {
+            None => return Ok(None),
+            Some(kv) => kv,
+        };
+
+        // please read todo! on implementation of .to_version()
+        let version = (&kv).to_version()?;
+        let (_, value) = kv.into_key_value();
+
+        if value.is_empty() {
+            // we keep an empty value in place of deleted
+            // keys to work around the version limitation
```

Comment:
outdated?

Reply:
Yes, definitely! The entire implementation can be improved now to not relay on the value being empty. Thanks for pointing this out